src/hotspot/share/opto/vector.cpp

```diff
@@ -314,182 +314,31 @@ void PhaseVector::scalarize_vbox_node(VectorBoxNode* vec_box) {
 
 void PhaseVector::expand_vbox_node(VectorBoxNode* vec_box) {
   if (vec_box->outcnt() > 0) {
-    Node* vbox = vec_box->get_oop();
-    Node* vect = vec_box->get_vec();
-    Node* result = expand_vbox_node_helper(vec_box, vbox, vect, vec_box->box_type(), vec_box->vec_type());
-    C->gvn_replace_by(vec_box, result);
-    C->print_method(PHASE_EXPAND_VBOX, 3, vec_box);
-  }
-  C->remove_macro_node(vec_box);
-}
+    Node* vbox_alloc = vec_box->get_oop();
+    assert(vbox_alloc->is_Proj() && vbox_alloc->in(0)->isa_VectorBoxAllocate(), "");
```

Comment:
Before we have the `PhiNode` as the oop and vect inputs, which are from the transformations like `merge_through_phi`, but now the oop is only the VBA?

Reply:
> Before we have the `PhiNode` as the oop and vect inputs, which are from the transformations like `merge_through_phi`, but now the oop is only the VBA?

I am banking on InlineTypeNode::merge_with to handle convergence cases instead of introducing special phi handling in box expansion.

Reply:
I wanted to get away with VectorBoxAllocateNode (a macro node) in a subsequent patch, I kept it currently because it being a SafePoint node carries JVMState for re-initialization of GraphKit during expansion, this helps in accessing control and memory edges needed during new_instance creation for vector_boxes.

Otherwise, InlineTypeNode handle control flow merges inherently.
https://github.com/openjdk/valhalla/blob/lworld%2Bvector/src/hotspot/share/opto/inlinetypenode.hpp#L91

What do you think ?

Reply:
**Issue around early VectorBoxAllocate expansions: -**

 
![image](https://user-images.githubusercontent.com/59989778/231967490-e0c699b7-d9c1-4998-916e-67fcfce543c1.png)


_Q. Why do we need an explicit  VectorBoxAllocate IR along with VectorBox?_

_A. This is because we need memory and control edges to stitch the graph pallet created after buffering, VectorBoxAllocate is a Safepoint node carrying JVMstate and other relevant edges thus re-initializing GraphKit corresponding to a given JVM state is seamless._ 

_Other possible solution could be to associate memory and control edges with VectorBox nodes OR bookkeep entire JVMState into VectorBox for a later use, but this will pose issues at convergence points where VectorBoxes coming from different control flow merge and so does current  VBAs and this is augmented by PhaseVector::expand_vbox_node_helper which processes each incoming VectorBoxAllocate separately._

_If we just preserve JVMState for a VectorBox then we need to have a synthetic IR for bookkeeping it since box forwarding across PhiNode result into creation of just one VectorBox at converge point, so keeping VBA looks relevant and clean solution for the time being._